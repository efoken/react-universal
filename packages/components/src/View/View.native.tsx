import {
  normalizeLayoutEvent,
  normalizeResponderEvent,
  normalizeRole,
  styled,
} from '@universal-ui/core';
import { forwardRef } from 'react';
import { View as RNView } from 'react-native';
import type { ViewProps, ViewType } from './View.types';

const ViewRoot = styled(RNView, {
  name: 'View',
  slot: 'Root',
})({
  position: 'static',
});

export const View = forwardRef<any, ViewProps>(
  (
    {
      onLayout,
      onMoveShouldSetResponder,
      onMoveShouldSetResponderCapture,
      onResponderEnd,
      onResponderGrant,
      onResponderMove,
      onResponderReject,
      onResponderRelease,
      onResponderStart,
      onResponderTerminate,
      onResponderTerminationRequest,
      onStartShouldSetResponder,
      onStartShouldSetResponderCapture,
      role,
      style,
      ...props
    },
    ref,
  ) => (
    <ViewRoot
      ref={ref}
      role={normalizeRole(role)}
      style={style as any}
      onMoveShouldSetResponder={normalizeResponderEvent(onMoveShouldSetResponder)}
      onMoveShouldSetResponderCapture={normalizeResponderEvent(onMoveShouldSetResponderCapture)}
      onResponderEnd={normalizeResponderEvent(onResponderEnd)}
      onResponderGrant={normalizeResponderEvent(onResponderGrant)}
      onResponderMove={normalizeResponderEvent(onResponderMove)}
      onResponderReject={normalizeResponderEvent(onResponderReject)}
      onResponderRelease={normalizeResponderEvent(onResponderRelease)}
      onResponderStart={normalizeResponderEvent(onResponderStart)}
      onResponderTerminate={normalizeResponderEvent(onResponderTerminate)}
      onResponderTerminationRequest={normalizeResponderEvent(onResponderTerminationRequest)}
      onStartShouldSetResponder={normalizeResponderEvent(onStartShouldSetResponder)}
      onStartShouldSetResponderCapture={normalizeResponderEvent(onStartShouldSetResponderCapture)}
      onLayout={normalizeLayoutEvent(onLayout)}
      {...props}
    />
  ),
) as unknown as ViewType;

View.displayName = 'View';
