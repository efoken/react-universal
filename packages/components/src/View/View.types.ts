import type { ResponderConfig } from '@tamagui/react-native-use-responder-events/types/ResponderSystem';
import type {
  LayoutEvent,
  RNStyle,
  StyleProp,
  SxProps,
} from '@universal-ui/core';
import type {
  NativeMethods,
  AccessibilityProps as RNAccessibilityProps,
  ViewProps as RNViewProps,
} from 'react-native';
import type { AccessibilityProps } from '../types';

export interface ViewMethods extends NativeMethods {}

export type ViewStyle = Omit<
  RNStyle,
  | 'color'
  | 'fontFamily'
  | 'fontSize'
  | 'fontStyle'
  | 'fontWeight'
  | 'letterSpacing'
  | 'lineHeight'
  | 'textAlign'
  | 'textDecorationColor'
  | 'textDecorationLine'
  | 'textDecorationStyle'
  | 'textTransform'
  | 'userSelect'
>;

export interface ViewProps
  extends Omit<
      RNViewProps,
      | keyof RNAccessibilityProps
      | 'focusable'
      | 'nativeID'
      | 'onLayout'
      | 'pointerEvents'
      | 'style'
    >,
    AccessibilityProps,
    Partial<
      Pick<
        ResponderConfig,
        | 'onScrollShouldSetResponder'
        | 'onScrollShouldSetResponderCapture'
        | 'onSelectionChangeShouldSetResponder'
        | 'onSelectionChangeShouldSetResponderCapture'
      >
    > {
  as?: React.ElementType;
  dir?: 'ltr' | 'rtl' | 'auto';
  href?: string;
  hrefAttrs?: {
    download?: any;
    rel?: string;
    target?: React.HTMLAttributeAnchorTarget;
  };
  lang?: Intl.UnicodeBCP47LocaleIdentifier;
  /**
   * Callback invoked on mount and layout changes.
   */
  onLayout?: (event: LayoutEvent) => void;
  style?: StyleProp<ViewStyle>;
  /**
   * The system prop that allows defining system overrides as well as additional
   * CSS styles.
   */
  sx?: SxProps;
}

export type ViewOwnerState = Pick<ViewProps, 'role'> & {
  hasTextAncestor: boolean;
};
