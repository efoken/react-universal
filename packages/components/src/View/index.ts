export { View } from './View';
export type { ViewMethods, ViewProps, ViewStyle } from './View.types';
