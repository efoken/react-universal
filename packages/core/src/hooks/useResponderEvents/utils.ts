const keyName = '__reactResponderId';

function isSelectionValid() {
  const selection = window.getSelection()!;
  const string = selection.toString();
  return (
    string.length > 0 &&
    string !== '\n' &&
    (selection.anchorNode?.nodeType === window.Node.TEXT_NODE ||
      selection.focusNode?.nodeType === window.Node.TEXT_NODE)
  );
}

function getEventPath(domEvent: Event): any[] {
  // The `selectionchange` event always has the document as the target. Use the
  // anchor node as the initial target to reconstruct a path. (We actually only
  // need the first responder node in practice.)
  if (domEvent.type === 'selectionchange') {
    const target = window.getSelection()?.anchorNode;
    return composedPathFallback(target);
  }
  return domEvent.composedPath != null
    ? domEvent.composedPath()
    : composedPathFallback(domEvent.target);
}

function composedPathFallback(target: EventTarget | Node | null | undefined): any[] {
  const path = [];
  while (target != null && target !== document.body) {
    path.push(target);
    target = (target as any).parentNode;
  }
  return path;
}

/**
 * Retrieve the responder ID from a host node.
 */
function getResponderId(node: any): number | undefined {
  if (node != null) {
    return node[keyName];
  }
}

/**
 * Store the responder ID on a host node.
 */
export function setResponderId(node: any, id: number) {
  if (node != null) {
    node[keyName] = id;
  }
}

/**
 * Filter the event path to contain only the nodes attached to the responder
 * system.
 */
export function getResponderPaths(domEvent: any): { idPath: number[]; nodePath: any[] } {
  const idPath: number[] = [];
  const nodePath = [];
  const eventPath = getEventPath(domEvent);
  for (const node of eventPath) {
    const id = getResponderId(node);
    if (id != null) {
      idPath.push(id);
      nodePath.push(node);
    }
  }
  return { idPath, nodePath };
}

/**
 * Walk the paths and find the first common ancestor.
 */
export function getLowestCommonAncestor(pathA: any[], pathB: any[]): any {
  let pathALength = pathA.length;
  let pathBLength = pathB.length;
  if (
    // If either path is empty
    pathALength === 0 ||
    pathBLength === 0 ||
    // If the last elements aren't the same there can't be a common ancestor
    // that is connected to the responder system
    pathA[pathALength - 1] !== pathB[pathBLength - 1]
  ) {
    return null;
  }

  let itemA = pathA[0];
  let indexA = 0;
  let itemB = pathB[0];
  let indexB = 0;

  // If A is deeper, skip indices that can't match.
  if (pathALength - pathBLength > 0) {
    indexA = pathALength - pathBLength;
    itemA = pathA[indexA];
    pathALength = pathBLength;
  }

  // If B is deeper, skip indices that can't match
  if (pathBLength - pathALength > 0) {
    indexB = pathBLength - pathALength;
    itemB = pathB[indexB];
    pathBLength = pathALength;
  }

  // Walk in lockstep until a match is found
  let depth = pathALength;
  while (depth--) {
    if (itemA === itemB) {
      return itemA;
    }
    itemA = pathA[indexA++];
    itemB = pathB[indexB++];
  }
  return null;
}

/**
 * Determine whether any of the active touches are within the current responder.
 * This cannot rely on W3C `targetTouches`, as Safari does not implement it.
 */
export function hasTargetTouches(target: any, touches: TouchList | undefined) {
  if (touches == null || touches.length === 0) {
    return false;
  }
  for (const touch of touches) {
    const node = touch.target;
    if (node != null) {
      if (target.contains(node)) {
        return true;
      }
    }
  }
  return false;
}

/**
 * Ignore `selectionchange` events that don't correspond with a person's intent
 * to select text.
 */
export function hasValidSelection(domEvent: Event) {
  if (domEvent.type === 'selectionchange') {
    return isSelectionValid();
  }
  return domEvent.type === 'select';
}

/**
 * Events are only valid if the primary button was used without specific
 * modifier keys.
 */
export function isPrimaryPointerDown(domEvent: any) {
  const touch = domEvent.type === 'touchstart' || domEvent.type === 'touchmove';
  const primaryMouseDown =
    domEvent.type === 'mousedown' && (domEvent.button === 0 || domEvent.buttons === 1);
  const primaryMouseMove = domEvent.type === 'mousemove' && domEvent.buttons === 1;
  const noModifiers = domEvent.altKey === false && domEvent.ctrlKey === false;

  if (touch || (primaryMouseDown && noModifiers) || (primaryMouseMove && noModifiers)) {
    return true;
  }
  return false;
}

export function isStartish(domEvent: Event): boolean {
  return domEvent.type === 'touchstart' || domEvent.type === 'mousedown';
}

export function isMoveish(domEvent: Event): boolean {
  return domEvent.type === 'touchmove' || domEvent.type === 'mousemove';
}

export function isEndish(domEvent: Event): boolean {
  return domEvent.type === 'touchend' || domEvent.type === 'mouseup' || isCancelish(domEvent);
}

export function isCancelish(domEvent: Event): boolean {
  return domEvent.type === 'touchcancel' || domEvent.type === 'dragstart';
}

export function isScroll(domEvent: Event): domEvent is Event {
  return domEvent.type === 'scroll';
}

export function isSelectionChange(domEvent: Event): domEvent is Event {
  return domEvent.type === 'select' || domEvent.type === 'selectionchange';
}
