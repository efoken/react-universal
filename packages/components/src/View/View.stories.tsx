import type { Meta, StoryObj } from '@storybook/react';
import { View } from './View';

const meta = {
  component: View,
} satisfies Meta<typeof View>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {};

export const PropAriaLabel: Story = {
  name: 'Prop "aria-label"',
  args: {
    'aria-label': 'accessibility label',
  },
};

export const PropAriaLabelledBy: Story = {
  name: 'Prop "aria-labelledby"',
  args: {
    'aria-labelledby': '123',
  },
};

export const PropAriaLive: Story = {
  name: 'Prop "aria-live"',
  args: {
    'aria-live': 'polite',
  },
};

export const PropRole: Story = {
  name: 'Prop "role"',
  args: {
    role: 'presentation',
  },
};

export const PropDir: Story = {
  name: 'Prop "dir"',
  args: {
    dir: 'ltr',
  },
};

export const PropHref: Story = {
  name: 'Prop "href"',
  args: {
    href: 'https://example.com',
  },
};

export const PropHrefAttrs: Story = {
  name: 'Prop "hrefAttrs"',
  args: {
    href: 'https://example.com',
    hrefAttrs: {
      download: 'filename.jpg',
      rel: 'nofollow',
      target: '_blank',
    },
  },
};

export const PropLang: Story = {
  name: 'Prop "lang"',
  args: {
    lang: 'fr',
  },
};

export const PropId: Story = {
  name: 'Prop "id"',
  args: {
    id: 'ID',
  },
};

export const PropStyle: Story = {
  name: 'Prop "style"',
  args: {
    style: { borderWidth: 5 },
  },
};

export const PropTestId: Story = {
  name: 'Prop "testID"',
  args: {
    testID: '123',
  },
};
