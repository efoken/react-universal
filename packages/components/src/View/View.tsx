'use client';

import { useResponderEvents } from '@tamagui/react-native-use-responder-events';
import type { AnyProps } from '@universal-ui/core';
import {
  forwardRef,
  forwardedProps,
  getLocaleDirection,
  styled,
  useElementLayout,
  useForkRef,
  useOwnerState,
  usePlatformMethods,
} from '@universal-ui/core';
import { isString, pick } from '@universal-ui/utils';
import { Children, useContext, useRef } from 'react';
import type { NativeMethods } from 'react-native';
import { TextAncestorContext } from '../Text/TextAncestorContext';
import type { ViewOwnerState, ViewProps } from './View.types';

function pickProps<T extends AnyProps>(props: T) {
  return pick(props, {
    ...forwardedProps.defaultProps,
    ...forwardedProps.accessibilityProps,
    ...forwardedProps.clickProps,
    ...forwardedProps.focusProps,
    ...forwardedProps.keyboardProps,
    ...forwardedProps.mouseProps,
    ...forwardedProps.touchProps,
    ...forwardedProps.styleProps,
    href: true,
    lang: true,
    onScroll: true,
    onWheel: true,
    pointerEvents: true,
  });
}

const ViewRoot = styled('div', {
  label: 'ViewRoot',
})<{ ownerState: ViewOwnerState }>(({ ownerState }) => ({
  alignItems: 'stretch',
  backgroundColor: 'transparent',
  borderColor: '#000',
  borderStyle: 'solid',
  borderWidth: 0,
  boxSizing: 'border-box',
  display: ownerState.hasTextAncestor ? ('inline-flex' as any) : 'flex',
  flexBasis: 'auto',
  flexDirection: 'column',
  flexShrink: 0,
  listStyleType: 'none',
  margin: 0,
  minHeight: 0,
  minWidth: 0,
  padding: 0,
  textDecorationLine: 'none',
}));

export const View = forwardRef<any, ViewProps>(
  (
    {
      as: asProp,
      hrefAttrs,
      onLayout,
      onMoveShouldSetResponder,
      onMoveShouldSetResponderCapture,
      onResponderEnd,
      onResponderGrant,
      onResponderMove,
      onResponderReject,
      onResponderRelease,
      onResponderStart,
      onResponderTerminate,
      onResponderTerminationRequest,
      onScrollShouldSetResponder,
      onScrollShouldSetResponderCapture,
      onSelectionChangeShouldSetResponder,
      onSelectionChangeShouldSetResponderCapture,
      onStartShouldSetResponder,
      onStartShouldSetResponderCapture,
      ...props
    },
    ref,
  ) => {
    if (process.env.NODE_ENV !== 'production') {
      for (const item of Children.toArray(props.children)) {
        if (isString(item)) {
          console.error(
            `Unexpected text node: ${item}. A text node cannot be a child of a <View>.`,
          );
        }
      }
    }

    const hasTextAncestor = useContext(TextAncestorContext);
    const hostRef = useRef<HTMLElement>(null);

    useElementLayout(hostRef, onLayout);
    useResponderEvents(hostRef, {
      onMoveShouldSetResponder,
      onMoveShouldSetResponderCapture,
      onResponderEnd,
      onResponderGrant,
      onResponderMove,
      onResponderReject,
      onResponderRelease,
      onResponderStart,
      onResponderTerminate,
      onResponderTerminationRequest,
      onScrollShouldSetResponder,
      onScrollShouldSetResponderCapture,
      onSelectionChangeShouldSetResponder,
      onSelectionChangeShouldSetResponderCapture,
      onStartShouldSetResponder,
      onStartShouldSetResponderCapture,
    } as any);

    let component: 'a' | 'div' = 'div';

    const langDirection =
      props.lang == null ? null : getLocaleDirection(props.lang);
    const componentDirection = props.dir ?? langDirection;

    const supportedProps: AnyProps = pickProps(props);
    supportedProps.dir = componentDirection;

    if (props.href != null) {
      component = 'a';
      if (hrefAttrs != null) {
        const { download, rel, target } = hrefAttrs;
        if (download != null) {
          supportedProps.download = download;
        }
        if (rel != null) {
          supportedProps.rel = rel;
        }
        if (isString(target)) {
          supportedProps.target = target.startsWith('_')
            ? target
            : `_${target}`;
        }
      }
    }

    const platformMethodsRef = usePlatformMethods(hostRef);
    const handleRef = useForkRef(hostRef, platformMethodsRef, ref);

    supportedProps.ref = handleRef;

    const ownerState = useOwnerState({
      hasTextAncestor,
      role: props.role,
    });

    return (
      <ViewRoot
        as={asProp ?? component}
        ownerState={ownerState}
        {...supportedProps}
      />
    );
  },
) as unknown as React.ComponentType<
  ViewProps & React.RefAttributes<NativeMethods>
> &
  NativeMethods;

View.displayName = 'View';
