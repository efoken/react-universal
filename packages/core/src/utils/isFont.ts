import { isObject } from '@react-universal/utils';
import type { AnyObject } from '@react-universal/utils';
import type { ThemeFont } from '../theme';

export function isFont(obj: AnyObject): obj is ThemeFont {
  return isObject(obj) && 'family' in obj && 'weights' in obj;
}
