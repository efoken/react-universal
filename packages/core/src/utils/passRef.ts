import { isFunction, noop } from '@react-universal/utils';

export function passRef<T>(
  value: T,
  ref: React.ForwardedRef<T>,
  onMount?: () => void,
  onUnmount?: () => void,
) {
  const passRef = () => {
    if (isFunction(ref)) {
      return ref(value);
    }
    if (ref != null) {
      ref.current = value;
    }
    return noop;
  };

  const refReturnFn = passRef();

  onMount?.();

  return () => {
    refReturnFn?.();
    onUnmount?.();
  };
}
