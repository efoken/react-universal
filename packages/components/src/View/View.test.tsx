import type { AnyObject } from '@react-universal/utils';
import { isFunction, noop } from '@react-universal/utils';
import { act, fireEvent, render } from '@testing-library/react';
import { createRef } from 'react';
import type { MockInstance } from 'vitest';
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import { View } from './View';
import type { ViewType } from './View.types';

function createEventTarget(node: any) {
  return {
    node,
    blur: () => fireEvent.blur(node),
    click: () => fireEvent.click(node),
    focus: (payload?: AnyObject) => fireEvent.focus(node, payload),
    pointerDown: (payload?: AnyObject) => fireEvent.pointerDown(node, payload),
  };
}

describe('View', () => {
  test('default', () => {
    const { container } = render(<View />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('non-text is rendered', () => {
    const children = <View testID="1" />;
    const { container } = render(<View>{children}</View>);
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('raw text nodes as children', () => {
    let consoleErrorMock: MockInstance<(...params: any[]) => void>;

    beforeEach(() => {
      consoleErrorMock = vi.spyOn(console, 'error').mockImplementation(noop);
    });

    afterEach(() => {
      consoleErrorMock.mockRestore();
    });

    test('error logged (single)', () => {
      render(<View>hello</View>);
      expect(console.error).toBeCalled();
    });

    test('error logged (array)', () => {
      render(
        <View>
          <View />
          hello
          <View />
        </View>,
      );
      expect(console.error).toBeCalled();
    });
  });

  describe('prop "aria-label"', () => {
    test('value is set', () => {
      const { container } = render(<View aria-label="accessibility label" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "aria-labelledby"', () => {
    test('value is set', () => {
      const { container } = render(<View aria-labelledby="123" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "aria-live"', () => {
    test('value is set', () => {
      const { container } = render(<View aria-live="polite" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "role"', () => {
    test('value is set', () => {
      const { container } = render(<View role="presentation" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('value is "button"', () => {
      const { container } = render(<View role="button" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('value alters HTML element', () => {
      const { container } = render(<View role="article" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "dir"', () => {
    test('value is "ltr"', () => {
      const { container } = render(<View dir="ltr" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('value is "rtl"', () => {
      const { container } = render(<View dir="rtl" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "href"', () => {
    test('value is set', () => {
      const { container } = render(<View href="https://example.com" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('href with role', () => {
      const { container } = render(<View role="presentation" href="https://example.com" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "hrefAttrs"', () => {
    test('requires "href"', () => {
      const { container } = render(<View hrefAttrs={{ download: 'filename.jpg' }} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('value is set', () => {
      const hrefAttrs = {
        download: 'filename.jpg',
        rel: 'nofollow',
        target: '_blank',
      };
      const { container } = render(<View href="https://example.com" hrefAttrs={hrefAttrs} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('target variant is set', () => {
      const hrefAttrs = {
        target: 'blank',
      };
      const { container } = render(<View href="https://example.com" hrefAttrs={hrefAttrs} />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('undefined values are excluded', () => {
      const hrefAttrs = {
        download: undefined,
        rel: undefined,
        target: undefined,
      };
      const { container } = render(<View href="https://example.com" hrefAttrs={hrefAttrs} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "lang"', () => {
    test('undefined', () => {
      const { container } = render(<View />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('fr', () => {
      const { container } = render(<View lang="fr" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('ar', () => {
      const { container } = render(<View lang="ar" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('with dir', () => {
      const { container } = render(<View dir="ltr" lang="ar" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "id"', () => {
    test('value is set', () => {
      const { container } = render(<View id="ID" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "onBlur"', () => {
    test('is called', () => {
      const onBlur = vi.fn();
      const ref = createRef<React.ElementRef<ViewType>>();
      act(() => {
        // @ts-expect-error: `onBlur` is Web only and does not exist in types
        render(<View ref={ref} onBlur={onBlur} />);
      });
      const target = createEventTarget(ref.current);
      act(() => {
        target.focus();
        target.blur();
      });
      expect(onBlur).toBeCalled();
    });
  });

  describe('prop "onClick"', () => {
    test('is called', () => {
      const onClick = vi.fn();
      const ref = createRef<React.ElementRef<ViewType>>();
      act(() => {
        // @ts-expect-error: `onClick` is Web only and does not exist in types
        render(<View ref={ref} onClick={onClick} />);
      });
      const target = createEventTarget(ref.current);
      act(() => {
        target.click();
      });
      expect(onClick).toBeCalled();
    });
  });

  describe('prop "onFocus"', () => {
    test('is called', () => {
      const onFocus = vi.fn();
      const ref = createRef<React.ElementRef<ViewType>>();
      act(() => {
        // @ts-expect-error: `onFocus` is Web only and does not exist in types
        render(<View ref={ref} onFocus={onFocus} />);
      });
      const target = createEventTarget(ref.current);
      act(() => {
        target.focus();
        target.blur();
      });
      expect(onFocus).toBeCalled();
    });
  });

  describe('prop "onPointerDown"', () => {
    // beforeEach(() => {
    //   setPointerEvent(true);
    // });
    // afterEach(() => {
    //   setPointerEvent(false);
    // });

    test('is called', () => {
      const onPointerDown = vi.fn();
      const ref = createRef<React.ElementRef<ViewType>>();
      act(() => {
        render(<View ref={ref} onPointerDown={onPointerDown} />);
      });
      const target = createEventTarget(ref.current);
      act(() => {
        target.pointerDown({ pointerType: 'touch' });
      });
      expect(onPointerDown).toBeCalled();
    });
  });

  describe('prop "ref"', () => {
    test('value is set', () => {
      const ref = vi.fn();
      render(<View ref={ref} />);
      expect(ref).toBeCalled();
    });

    test('is not called for prop changes', () => {
      const ref = vi.fn();
      let rerender: (ui: React.ReactNode) => void;
      act(() => {
        ({ rerender } = render(<View ref={ref} id="123" style={{ borderWidth: 5 }} />));
      });
      expect(ref).toHaveBeenCalledTimes(1);
      act(() => {
        rerender(<View ref={ref} id="1234" style={{ borderWidth: 6 }} />);
      });
      expect(ref).toHaveBeenCalledTimes(1);
    });

    test('node has imperative methods', () => {
      const ref = createRef<React.ElementRef<ViewType>>();
      act(() => {
        render(<View ref={ref} />);
      });
      const node = ref.current!;
      expect(isFunction(node.measure));
      expect(isFunction(node.measureLayout));
      expect(isFunction(node.measureInWindow));
    });
  });

  describe('prop "style"', () => {
    test('value is set', () => {
      const { container } = render(<View style={{ borderWidth: 5 }} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('prop "testID"', () => {
    test('value is set', () => {
      const { container } = render(<View testID="123" />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
