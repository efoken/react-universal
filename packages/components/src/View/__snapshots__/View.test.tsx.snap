// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`View > default 1`] = `
<div
  class="css-1uo959j"
/>
`;

exports[`View > non-text is rendered 1`] = `
<div
  class="css-1uo959j"
>
  <div
    class="css-1uo959j"
    data-testid="1"
  />
</div>
`;

exports[`View > prop "aria-label" > value is set 1`] = `
<div
  aria-label="accessibility label"
  class="css-1uo959j"
/>
`;

exports[`View > prop "aria-labelledby" > value is set 1`] = `
<div
  aria-labelledby="123"
  class="css-1uo959j"
/>
`;

exports[`View > prop "aria-live" > value is set 1`] = `
<div
  aria-live="polite"
  class="css-1uo959j"
/>
`;

exports[`View > prop "dir" > value is "ltr" 1`] = `
<div
  class="css-1uo959j"
  dir="ltr"
/>
`;

exports[`View > prop "dir" > value is "rtl" 1`] = `
<div
  class="css-1uo959j"
  dir="rtl"
/>
`;

exports[`View > prop "href" > href with role 1`] = `
<a
  class="css-1uo959j"
  href="https://example.com"
  role="presentation"
/>
`;

exports[`View > prop "href" > value is set 1`] = `
<a
  class="css-1uo959j"
  href="https://example.com"
/>
`;

exports[`View > prop "hrefAttrs" > requires "href" 1`] = `
<div
  class="css-1uo959j"
/>
`;

exports[`View > prop "hrefAttrs" > target variant is set 1`] = `
<a
  class="css-1uo959j"
  href="https://example.com"
  target="_blank"
/>
`;

exports[`View > prop "hrefAttrs" > undefined values are excluded 1`] = `
<a
  class="css-1uo959j"
  href="https://example.com"
/>
`;

exports[`View > prop "hrefAttrs" > value is set 1`] = `
<a
  class="css-1uo959j"
  download="filename.jpg"
  href="https://example.com"
  rel="nofollow"
  target="_blank"
/>
`;

exports[`View > prop "id" > value is set 1`] = `
<div
  class="css-1uo959j"
  id="ID"
/>
`;

exports[`View > prop "lang" > ar 1`] = `
<div
  class="css-1uo959j"
  dir="rtl"
  lang="ar"
/>
`;

exports[`View > prop "lang" > fr 1`] = `
<div
  class="css-1uo959j"
  dir="ltr"
  lang="fr"
/>
`;

exports[`View > prop "lang" > undefined 1`] = `
<div
  class="css-1uo959j"
/>
`;

exports[`View > prop "lang" > with dir 1`] = `
<div
  class="css-1uo959j"
  dir="ltr"
  lang="ar"
/>
`;

exports[`View > prop "role" > value alters HTML element 1`] = `
<article
  class="css-1uo959j"
/>
`;

exports[`View > prop "role" > value is "button" 1`] = `
<button
  class="css-1uo959j"
  type="button"
/>
`;

exports[`View > prop "role" > value is set 1`] = `
<div
  class="css-1uo959j"
  role="presentation"
/>
`;

exports[`View > prop "style" > value is set 1`] = `
<div
  class="css-1uo959j"
  style="border-width: 5px;"
/>
`;

exports[`View > prop "testID" > value is set 1`] = `
<div
  class="css-1uo959j"
  data-testid="123"
/>
`;
